ethcore/blockchain/src/database_extras.rs

```diff
@@ -14,14 +14,27 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
 
-//! Bridge between Tracedb and Blockchain.
+//! Provides a `DatabaseExtras` trait that defines an interface to query for block data not
+//! contained in a TraceDB.
 
-use blockchain::{BlockChain, BlockProvider, TransactionAddress};
+use common_types::BlockNumber;
 use ethereum_types::H256;
-use trace::DatabaseExtras as TraceDatabaseExtras;
-use types::BlockNumber;
+use ethcore_db::keys::TransactionAddress;
 
-impl TraceDatabaseExtras for BlockChain {
+use crate::blockchain::{BlockProvider, BlockChain};
+
+/// `DatabaseExtras` provides an interface to query extra data which is not stored in TraceDB,
+/// but necessary to work correctly.
+pub trait DatabaseExtras {
```

Comment:
To be fair IMHO this trait should be part of the `TraceDB` modules, cause it seems it's required there.
IMHO any place where we have a trait and an implementation for a non-dummy struct right below suggests some issue with code split.

Reply:
@debris asked for it to be moved here to avoid tight coupling. Personally I think it is better off here, but if you feel strongly otherwise I can move it back.

Reply:
Well now, we couple `trace` with `blockchain`. I think it would be best to have the trait in `trace` but keep the implementation outside of `blockchain` crate as well (the impl should live in a place where it's actually instantiated) - that way you can avoid coupling at all.

Reply:
> @debris asked for it to be moved here to avoid tight coupling. Personally I think it is better off here, but if you feel strongly otherwise I can move it back.

not exactly. I wanted this trait to remain in trace library, but only an implementation of it to be moved elsewhere. Something like this:

- blockchain - defines `BlockchainDB`
- trace - defines `TraceDB` and `DatabaseExtras`
- ethcore - defines a structure that holds `BlockchainDB` (e.g. Client) and implements `DatabaseExtras` for it

Reply:
Addressed [here](https://github.com/paritytech/parity-ethereum/pull/10868).